tools/cldr-code/src/main/java/org/unicode/cldr/util/VettingViewer.java

```diff
@@ -900,17 +920,103 @@ protected void compute() {
+                            System.err.println("START: " + a);
+                            a.computeAll();
+                            System.err.println("DONE: " + a);
+                        }
+                    });
+                    t.start();
+                    threads.add(t);
+                    i += chunk;
+                }
+                System.err.println("VettingViewer: spun up " + threads.size() + " thread(s).");
+                // now wait for the threads to finish
+                if (DEBUG_THREADS) {
+                    int liveCount = threads.size();
+                    while(!progressCallback.isStopped() && liveCount > 0) {
+                        liveCount = 0;
+                        for(final Thread t : threads) {
+                            if (t.isAlive()) {
+                                liveCount++;
+                            }
+                        }
+                        if (DEBUG_THREADS) {
+                            System.err.println("THREAD COUNT: " + liveCount+"/"+threads.size() + " alive");
+                        }
+                        if(liveCount > 0) {
+                            try {
+                                Thread.sleep(1000);
+                            } catch(InterruptedException e) {
+
+                            }
+                        }
+                    }
+                }
+                // just wait for all threads to finish
+                for (Thread t : threads) {
+                    t.join();
+                }
+                if (DEBUG_THREADS) {
+                    System.err.println("All threads done.");
+                }
+            }
+        }
+
         /**
          * Compute this entire task.
          * Can call this to run this step as a single thread.
          */
         private void computeAll() {
+            if (DEBUG_THREADS) {
+                System.out.println("ComputeAll: " + start + ".."+(start+length));
+            }
             // do this many at once
             for (int n = start; n < (start + length); n++) {
                 computeOne(n);
             }
         }
 
+        private static final long MEMORY_LOW_WATER_MARK = 64000000L; // TODO: Parameterize
+
+        private void waitForFreeMemory() {
+            final Runtime r = Runtime.getRuntime();
+            long freeMem = r.freeMemory();
+            while (freeMem < MEMORY_LOW_WATER_MARK) {
+                System.err.println(this + " - sleeping because free mem " + freeMem + " below " + MEMORY_LOW_WATER_MARK);
+                try {
+                    Thread.sleep(10000); // TODO: randomize
+                } catch (InterruptedException ie) {
+                    throw new OutOfMemoryError(this + " InterruptedException while waiting for free space");
+                }
+                freeMem = r.freeMemory();
+            }
+            System.err.println(this + " - memory " + freeMem + " OK");
+        }
```

Comment:
Both VettingViewer and VettingViewerQueue could be improved by more separation of concerns. Some concerns are CLDR-specific: locales, xpaths, votes, organizations, ...

Other concerns are generic: threads, memory, progress bars, ...

waitForFreeMemory is generic and potentially reusable. It will be more at home in a class (not in VettingViewer) that knows about threads and memory but knows nothing CLDR-specific.